packages/labs/ssr/src/lib/lit-element-renderer.ts

```diff
@@ -66,14 +66,19 @@ export class LitElementRenderer extends ElementRenderer {
   }
 
   override connectedCallback() {
+    // Currently we only call connectedCallback as an opt-in,
+    // as this could break consumers.
+    if (globalThis.litSsrCallConnectedCallback) {
+      this.element.connectedCallback();
```

Comment:
I'm not clear on why this is ok here. Doesn't it call `createRenderRoot` and enable asynchronous updating? Both seem problematic...

Reply:
What would you suggest? Wrapping the call to `createRenderRoot` and `enableUpdating` in an if with `isServer`?
A separate method (e.g. `connectedCallbackServer`) was rejected during a previous community call.

Reply:
Try `LitElement.prototype.connectedCallback = function() {}` and probably should add a test that confirms the element does *not* try to update (e.g. implement firstUpdated and wait for a setTimeout and confirm it's not called).

Reply:
My understanding was that we want to call the controllers (`this.__controllers?.forEach((c) => c.hostConnected?.());`), in order for the context package to work. So should it be `LitElement.prototype.connectedCallback = function() { this.__controllers?.forEach((c) => c.hostConnected?.()) }`?
Also where should I patch this? The first time it is called in `LitElementRenderer`?

Reply:
Yeah, that's fair. Err, `__controllers` would likely have to be added to `lit-element/private-ssr-support.js`.

Perhaps instead, we could no-op `createRenderRoot` and `enableUpdating` but `enableUpdating` is set on the instance so it's probably simplest just to add this before calling `connectedCallback()`:

```js
// Optionally call connectedCallback via setting: `litSsrCallConnectedCallback`
// Enable this flag to process events dispatched handled via connectedCallback.
if (globalThis.litSsrCallConnectedCallback) {
  // Prevent element from trying to create a shadowRoot or enable updating
  // since these are both specially handled for server rendering.
  this.element.createRenderRoot = function() {};
  this.element.enableUpdating = function() {};
  this.element.connectedCallback();
}
// ...
```

And we should add tests to confirm this behavior.

Reply:
Quick question; With this the controller hook `hostDisconnected` is called, but the `hostUpdate` hook, which is nested between `willUpdate` and `update`, will not be called. Is that acceptable for now?

Reply:
I assume you mean `hostConnected`? 

I'm not sure about `hostUpdate`. I think we can start with not calling it and just making sure it's commented clearly that it's not called.

Reply:
Yes, I meant `hostConnected`. 😅 
How should I best document this? In the `ReactiveController` docs?

Reply:
In the code where `willUpdate` is specially called and perhaps add a bullet [here](https://github.com/lit/lit/tree/main/packages/labs/ssr#notes-and-limitations) about events and controllers?

Reply:
Done so. Can you have another look?

Additional question for `@lit/context`; In the browser you can use e.g. `document.body` for `ContextRoot`. It would be helpful if we had something similar for the server. However, currently my only idea is to define a constant on `globalThis` (e.g. `serverEventRoot`), which would be added to the start of `RenderInfo.eventTargetStack`.
Then you could use `new ContextRoot().attach(isServer ? serverEventRoot : document.body)`.
What do you think? Do you have another idea?

Reply:
So don't touch `createRenderRoot` 👍 

As for `enqueueUpdate`; How so? It is a private method? Or what do you mean?

Reply:
* `createRenderRoot` calls `adoptStyles` to add styles to the `shadowRoot`. Calling this is either bad or unncessary.
* It would probably be easiest to no-op `enableUpdating`.